pkg/controllers/grafana-operator/controller.go

```diff
@@ -279,50 +279,62 @@ func (r *reconciler) reconcileSubscription(ctx context.Context) reconcileResult
 	}
 
 	// create
+	desired := NewSubscription()
 	if errors.IsNotFound(err) {
 		r.logger.Info("Creating Grafana Operator Subscription")
-		err := r.k8sClient.Create(ctx, NewSubscription())
+		err := r.k8sClient.Create(ctx, desired)
 		return creationResult(err)
 	}
 
 	r.logger.Info("Updating Grafana Operator Subscription")
-	subscription.Spec = NewSubscription().Spec
-	err = r.k8sClient.Update(ctx, &subscription)
-	return updationResult(err)
+	subscription.Spec.StartingCSV = desired.Spec.StartingCSV
+	return updationResult(r.k8sClient.Update(ctx, &subscription))
 }
 
 func (r *reconciler) approveInstallPlan(ctx context.Context) reconcileResult {
-	subscriptionKey := types.NamespacedName{
-		Name:      subscriptionName,
-		Namespace: Namespace,
-	}
-	var subscription v1alpha1.Subscription
-	if err := r.k8sClient.Get(ctx, subscriptionKey, &subscription); err != nil {
+
+	var installPlans v1alpha1.InstallPlanList
+	err := r.k8sClient.List(ctx, &installPlans, client.InNamespace(Namespace))
+	if err != nil {
 		return reconcileError(err)
 	}
 
-	installPlanRef := subscription.Status.InstallPlanRef
-	if installPlanRef == nil {
+	// wait for install plans to be created
+	if len(installPlans.Items) == 0 {
 		return end()
 	}
 
-	installPlanKey := types.NamespacedName{
-		Namespace: Namespace,
-		Name:      installPlanRef.Name,
-	}
-	var installPlan v1alpha1.InstallPlan
-	err := r.k8sClient.Get(ctx, installPlanKey, &installPlan)
-	if err != nil {
-		return reconcileError(err)
+	var approvePlan *v1alpha1.InstallPlan
+
+	for _, installPlan := range installPlans.Items {
+		csv := installPlan.Spec.ClusterServiceVersionNames[0]
+		// ignore all but the install matching the Grafana version
+		// also ignore install-plans that has an empty status
+		if csv != grafanaCSV || len(installPlan.Status.BundleLookups) == 0 {
+			continue
+		}
+
+		r.logger.V(8).Info("install plan", "name", installPlan.Name, "csv", csv, "approved", installPlan.Spec.Approved)
+
+		// look no further if an install plan for the desired CSV is already approved
+		if installPlan.Spec.Approved {
+			r.logger.V(8).Info("install plan already approved", "name", installPlan.Name, "csv", csv)
+			return next()
+		}
+
+		approvePlan = &installPlan
```

Comment:
Do we want to approve all install plans or just one?

Reply:
There should be only one plan if we wait for the plan to be properly initialized.

Reply:
Ok, sounds good :+1: